### ISSUE TITLE
Validate the eval material before trusting any CER: score a known-good model on GT_Thun-Test

### ISSUE THREAD

**Issue by thodel** [thodel/serving-atr-inference#52]
Every CER this subsystem has produced was measured on the same eval material, and
all of them are **insertion-dominated**. Before another training run is tuned
against those numbers, the numbers themselves need one check.

## The measurements

Both kraken runs were scored on `GT_Thun-Test_(DEMO_TEST)` — identical `chars`,
so identical material:

| run | chars | errors | CER | insertions | deletions | substitutions |
|---|---|---|---|---|---|---|
| `kraken-thun-missiven-v1` (`20260807T161137Z`) | 11,566 | 11,379 | **0.9838** | 11,191 | 2 | 186 |
| `kraken-medieval-scripts-v1` (`20260807T212539Z`) | 11,566 | 8,182 | **0.7074** | 5,381 | 48 | 2,753 |

The VLM smoke run lands in the same place from a different architecture: CER
**0.466** adapted, **1.837** for the un-adapted base — a CER above 1 is only
reachable by emitting far more text than the reference contains.

Deletions are 2 and 48. Every model tried here, CTC and autoregressive alike,
**outputs substantially more characters than the reference says are on the line.**

## Two readings, opposite fixes

1. **The training design.** `kraken-medieval-scripts-v1` trained on one small Thun
   project against seven large Leuven ones (`Itinera_Nova_100pages`,
   `SAL7305`–`SAL7370`) and was scored purely on Bernese transfer. A model shaped
   by Flemish hands scoring 29 % on Thun is an unsurprising result, and the fix is
   in the job: train Thun-only, or fine-tune from a Leuven base, or evaluate on a
   held-out slice of the same mixture.
2. **The eval material.** If line crops are paired with references that are short
   or offset — a `TextEquiv` that holds part of the line, a `Coords` polygon
   covering more than its transcription — then a model reading the image correctly
   scores as insertions, and *every* number above is measuring the ground truth
   rather than the model.

Validation cannot separate them: `best_0.2925.mlmodel` agrees with the test's
29.26 % to four digits, but training validation and test use the same data, so
consistency proves only that both stages see the same thing.

## The check

**Step 1 — look at five pairs (minutes).** Take five line crops from a compiled
eval set and read them against their reference text. If the text does not match
what is in the image, reading (2) is confirmed and nothing else in this issue
matters. The VLM job's `eval_report.json` already holds generated-vs-reference
pairs and is the cheapest window; the kraken side needs a crop dumped from the
`.arrow`.

**Step 2 — score a known-good model on the same material.** Pick a published
Zenodo kraken model from `config/models.yaml` that suits XIV–XVI German
chancery hands, and run it over `GT_Thun-Test` through `eval/run_eval.py`.

| result | verdict | what follows |
|---|---|---|
| CER ≈ 0.05–0.15 | the eval material is sound | reading (1): the training runs are poor, tune the job |
| CER ≈ 0.5–0.9, insertion-dominated | the material is broken | reading (2): fix `prepare`/`compile` pairing before any further training |

A published model that has never seen this corpus is the control. It is the only
measurement available that does not depend on our own pipeline being right.

## Why this outranks the rest of the training board

#36, #38, #40, #45 and #46 all make the subsystem better at producing numbers.
This one establishes whether the numbers mean anything. Until it runs, no CER
here should be quoted, published, or used to compare two runs — and
`scripts/publish_to_hub.py` is deliberately blocked on exactly that judgement,
since a model card's headline metric is a claim.

Extends **#37** (M5: runbook + eval comparison), which frames the same harness as
old-vs-new comparison; this is the prior question of whether the harness'
reference material is trustworthy at all.

## Acceptance

- Five inspected crop/reference pairs, recorded in the issue.
- An `eval/run_eval.py` report for one known-good registry model on `GT_Thun-Test`.
- A verdict written into `docs/TRAINING_PLAN.md`, so the next reader of a CER in
  this repo knows what it rests on.

**Comment by thodel** [comment]
Concrete evidence for the "alien eval set" branch, from the currently running `20260808T183111Z-kraken-medieval-full-v1`:

```
train_projects:  GT_Thun-Training_(TEST-DEMO), Itinera_Nova_100pages,
                 and ~690 SAL* projects                       → Leuven / Itinera Nova, Flemish
eval_projects:   GT_Thun-Test_(DEMO_TEST)                     → Bernese German
                 SAL7301_full, SAL8350_full                   → Flemish
                 u-17_0187a                                   → ?
max_pages: null, partition: 0.9, epochs: 5
```

The training material is overwhelmingly Flemish Leuven; the eval set mixes Bernese German into it. **One corpus-level CER over that union is not a result on either corpus** — it is an average across two scripts, two languages and two periods, weighted by whatever the page counts happen to be.

That matters for reading the earlier runs too:

| run | trained on | evaluated on | CER | ins/del |
|---|---|---|---|---|
| `kraken-thun-missiven-v1` | Thun | Thun | 0.9838 | 11,191 / 2 |
| `kraken-medieval-scripts-v1` | mixed | mixed | 0.7074 | 5,381 / 48 |

**The mismatch theory does not explain the first one.** Thun trained, Thun evaluated, and still 11,191 insertions against 2 deletions. A corpus mismatch cannot produce that asymmetry — the model emitted roughly 11 K characters more than the reference contained, on material of its own kind.

So there are two distinguishable problems here, and this issue should separate them:

1. **A mixed eval set makes any single CER uninterpretable** — real, visible in the job spec above, and fixable by scoring each corpus separately (relates to #55, which asks for the same decomposition on the metric side).
2. **Something makes every model over-generate**, independent of corpus. That is what scoring a published Zenodo model on `GT_Thun-Test` settles: if a model that never saw this material *also* over-generates on it, the eval material is at fault; if it scores normally, our training or compile step is.

Only the second needs the Zenodo model. The first can be checked immediately by evaluating an existing trained model against each eval project on its own — `ketos test` takes a manifest, so it is four runs of an existing artifact, no retraining.

**Comment by thodel** [comment]
## The eval material is not the problem — measured

`scripts/audit_eval_material.py` on the Thun eval split (c7e31a5, ef7c10e):

```
pages           12        lines 189        characters 11,502
chars/line      mean 60.9   p5 10   p50 66   p95 86
line width px   p5 148      p50 806          p95 961
px per char     mean 13.5   p5 10   p50 12.15  p95 17.65   (plausible 6–60)
implausible     3 lines (1.6%)

PLAUSIBLE — median 12.15 px per character over 189 lines, 1.6% outside 6–60.
```

Two of the three flagged lines turned out to be **vertical** (64×310 px, 51×399 px — marginalia or rotated regions), which the heuristic was measuring across rather than along; fixed in ef7c10e. The third is a genuine two-character line fragment. So the references are the right length for their crops, and **no branch of this issue that blames the ground truth survives**.

## What that leaves, and the evidence for it

`kraken-thun-missiven-v1` trained **from scratch** on this material: 139 pages, 2,087 lines, `partition: 0.9` → roughly **1,900 training lines**, into the 15.2 M-parameter `kraken+` network with no `--load`.

That is one to two orders of magnitude short of what a CTC network needs from random initialisation, and it predicts the exact signature observed:

- An **unconverged** CTC model has not yet learned blank-dominance. Its per-timestep softmax is near-uniform, so greedy decoding emits a character at almost every timestep.
- The kraken+ spec downsamples width by 8, so an 806 px line (the median here) yields roughly 100 timesteps — against a 66-character reference.
- Emitting ~100 characters where 66 are expected produces **insertions and almost no deletions**: 11,191 vs 2, over 11,502 reference characters. The hypothesis is about twice the reference length, which is what ~100 timesteps against ~66 characters gives.

A *well-trained* CTC model cannot over-generate this way, and a *collapsed* one emits blanks and scores deletions. Only the middle state — trained, but nowhere near enough — produces this.

`kraken-medieval-scripts-v1` fits the same curve from the other side: more data, CER 0.707 instead of 0.984, insertions still dominant (5,381 vs 48). Less unconverged, same regime.

## So the answer to this issue is: neither the material nor a bug — the runs were under-resourced

Which makes the Zenodo comparison **no longer the decisive experiment**. It would confirm the material is readable, which the audit now shows more cheaply. Two better next steps:

1. **Fine-tune instead of training from scratch on small corpora.** `TrainRequest.base_model` already accepts a registry id or a Zenodo DOI and produces `ketos train --load … --resize union` (`docs/TRAINING_PLAN.md` §4, `_resolve_base_model`). 1,900 lines is a perfectly reasonable *fine-tuning* set and a hopeless *from-scratch* one. This is an hour's run, not a day's.
2. **Let the full medieval run be the test of the other direction.** `20260808T183111Z-kraken-medieval-full-v1` is from-scratch on ~18 M lines. If this analysis is right it should converge and produce a defensible CER; if it also over-generates, the explanation is wrong and something is broken in compile or decoding after all.

## What should still be built here

The audit belongs in the pipeline rather than in a script someone remembers to run — the prepare stage already has the PageXML in hand and could record the px/char distribution on the job record, so a job that materializes implausible material says so before spending a GPU-day on it. Suggest that as the remaining scope of this issue, with the diagnosis above closing the original question.

**Comment by thodel** [comment]
## Confirmed, and the number is smaller than "too little data" suggests

```
base_model: null        # from scratch
epochs:     50
pages: 139   lines: 2,087   partition 0.9  →  ~1,878 training lines
batch_size: 256
```

**1,878 ÷ 256 = 7 steps per epoch. × 50 epochs = ~367 optimizer steps.**

A 15.2 M-parameter network from random initialisation received 367 gradient updates. For scale, the VLM smoke run took 38 steps and produced fluent-looking invented text; 367 on a from-scratch CTC network is barely past initialisation, which is exactly the regime that emits a character at nearly every timestep and scores as insertions.

`--schedule 1cycle` makes it worse: kraken derives the cycle length from `--epochs` and steps `OneCycleLR` per batch, so the learning rate ramps and anneals across those 367 steps. The run never reaches a productive learning rate at all.

So the batch size is the proximate cause. 256 is right for the 18 M-line corpus these defaults were written for and wrong by three orders of magnitude for 1,878 lines.

## The fix, and a prediction

Two changes, either of which alone helps, both together give a real result in under an hour:

```json
{
  "engine": "kraken",
  "model_id": "kraken-thun-finetune-v1",
  "base_model": "kraken-medieval_generic_b",
  "dataset": { "...as before..." },
  "params": { "batch_size": 16, "epochs": 30, "resize": "union" }
}
```

- **`base_model`** turns this from "train a network" into "adapt one", which is what 1,878 lines can support. `_resolve_base_model` fetches the Zenodo DOI through htrmopo, `ketos_cmd.train_cmd` emits `--load … --resize union` and correctly omits `--spec` (kraken ignores it when loading). `kraken-medieval_generic_b` is 14th–16th c. medieval script; `union` extends its Latin codec with the German characters Thun needs.
- **`batch_size: 16`** gives 117 steps/epoch → ~3,500 steps over 30 epochs, roughly 10× more updates than the failed run even before counting the head start from the base model.

**Prediction, stated before the fact:** `20260808T183111Z-kraken-medieval-full-v1` is from-scratch but has ~18 M lines → ~72 K steps/epoch × 5 = **~360 K steps**. If this analysis is right, that run should converge and produce a defensible CER despite being from scratch, because it is the corpus size the defaults assume. If it *also* comes back insertion-dominated, this explanation is wrong and the problem is in compile or decoding.

## Remaining scope for this issue

The diagnosis closes the original question — the material is fine (audited), the runs were under-resourced. What is worth keeping:

1. **Warn at submit when the step count is implausibly low.** `lines / batch_size × epochs` is computable from the request the moment `prepare` reports its line count, and "this configuration will take 367 optimizer steps" is a sentence that would have saved two runs and two days. It belongs next to the disk and VRAM guards.
2. **Record the px/char audit on the job record** during prepare, so material problems surface without anyone remembering to run a script.

**Comment by thodel** [comment]
Settled — and by the cheaper of the two routes this issue proposed.

**The eval material is sound.** `scripts/audit_eval_material.py` on the Thun split reports a median of **12.15 px of line per reference character** over 189 lines, with **98.4 %** inside the plausible band. The two apparent outliers were vertical lines the first heuristic measured across instead of along (fixed in ef7c10e). So the second hypothesis in this issue — crops paired with short or offset references — is ruled out.

**The cause is the training configuration**, and the arithmetic is unambiguous: `kraken-thun-missiven-v1` trained *from scratch* on ~1,878 lines at `batch_size: 256` — 7 optimizer steps per epoch, ~367 over 50 epochs, for a 15.2 M-parameter network from random weights, with `1cycle` ramping and annealing across all of them. An unconverged CTC network has not learned blank-dominance and emits a character at nearly every timestep, which *is* an insertion-dominated CER. The defaults are correct for the ~18 M-line corpus they were written against and wrong by three orders of magnitude for 1,878 lines. Recorded in 1d18f2a, README §"the first three runs were under-configured" and TRAINING_PLAN §9a.

**The expensive half is no longer needed.** Scoring a published Zenodo model as a control would have distinguished the same two hypotheses; the audit distinguished them without a GPU, a download or a run.

**What came out of it, which is the part that lasts:**
- `training/eval_material.py` + `scripts/audit_eval_material.py` — run it on any job before believing its CER.
- **#72**, the step-count guard: the configuration that caused this is now refused between prepare and compile, before any GPU time, with the arithmetic and three remedies in the error. That closes the loop this issue opened — the question was "can these numbers be trusted", and the answer is now enforced rather than remembered.

PR #67 proposed the same audit and was closed as a duplicate of the commits already on main.

**Comment by thodel** [comment]
**Correction to the arithmetic I posted above** (fixed in 214f49e). The diagnosis is unchanged; two of the numbers in it were wrong.

| | I wrote | actually |
|---|---|---|
| training lines | ~1,878 | **1,898** |
| optimizer steps | ~367 | **400** |

- **Lines:** I derived 1,878 from `partition: 0.9`. That job used explicit `eval_projects`, so the partition never applied — the split is whatever the eval projects contain. The material audit counted the eval side directly at **189 lines**, so against 2,087 transcribed the training set is 1,898.
- **Steps:** 1,898 ÷ 256 = 7.4, which is **8** batches per epoch (7 full + 1 partial), so 400 over 50 epochs. My 367 came from multiplying the fraction through instead of rounding per epoch — neither the floor (350) nor the ceiling (400) of the real figure.

Worth correcting because the step-count guard (#72) repeats it to operators as a measured fact: *"it was 400 steps that produced CER 0.98 on the Thun set."* It should be the measured fact.

Everything downstream is unaffected — 400 against a floor of 2,000 is the same verdict, and the mechanism (an unconverged CTC network emitting a character at nearly every timestep) does not turn on 350 vs 400.
